File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/DotNet/PowerShell/VollistCmdlet/Vollist.cs
//==================================================================//
//                                                                  //
// $Id: $                                                           //
// Vollist.cs                                                       //
//                                                                  //
// This sample code demonstrates how to create a cmdlet for         //
// listing NetApp volumes.                                          //
//                                                                  //
// Copyright 2009 NetApp. All rights reserved. Specifications       //
// subject to change without notice.                                //
//                                                                  //
// This SDK sample code is provided AS IS, with no support or       //
// warranties of any kind, including but not limited to             //
// warranties of merchantability or fitness of any kind,            //
// expressed or implied.  This code is subject to the license       //
// agreement that accompanies the SDK.                              //
//                                                                  //
// Usage: Get-Volume -server <server> -user <user> -passwd          //    
//    <password> [-name <volume>]                                   //
//                                                                  //
// Compile the project which will generate vollist.dll.             //
// Perform the vollist.dll installation with .NET using             //
//    InstallUtil.exe command.                                      //
// Add the snapin to PowerShell using the command:                  //
//      add-pssnapin NetApp.ManageOntap.Volume                      //
// To remove the snapin, do so with this command:                   //
//      remove-pssnapin NetApp.ManageOntap.Volume                   //
//  To execute, issue this command:                                 //
//      Get-Volume -server <server> -user <user> -passwd            //
//          <password> [-name <volume>]                             //
//==================================================================//

using System;
using System.Collections.Generic;
using System.Text;
using System.Management.Automation;
using System.ComponentModel;
using NetApp.Manage;

namespace NetApp.ManageOntap.CmdLets.Volume
{
    // Get-Volume Cmdlet - This cmdlet will list the 
    // volumes on a Filer.
    [Cmdlet(VerbsCommon.Get, "Volume")]
    public class GetVolume : PSCmdlet
    {
        private string _server = null;
        private string _user = null;
        private string _passwd = null;
        private string _volume = null;

        [Parameter(
            Position = 0,
            HelpMessage = "Name of the Server")]
        public string Server
        {
            get { return _server; }
            set { _server = value; }
        }

        [Parameter(
            Position = 0,
            HelpMessage = "User Name")]
        public string User
        {
            get { return _user; }
            set { _user = value; }
        }

        [Parameter(
            Position = 0,
            HelpMessage = "Password")]
        public string Passwd
        {
            get { return _passwd; }
            set { _passwd = value; }
        }

        [Parameter(
            Position = 0,
            HelpMessage = "Name of the volume")]
        public string Name
        {
            get { return _volume; }
            set { _volume = value; }
        }

        public String Usage
        {
            get { return "Usage: Get-Volume -server <server> -user <user> -passwd <password> [-name <volume>]"; }
        }

        protected override void ProcessRecord()
        {
            if (_server == null || _user == null || _passwd == null)
            {
                WriteObject(Usage);
                return;
            }

            try
            {
                NaServer server = new NaServer(_server, 1, 0);
                NaElement input = new NaElement("volume-list-info");
                server.SetAdminUser(User, Passwd);
                if (_volume != null)
                {
                    input.AddNewChild("volume", _volume);
                }
                NaElement output = server.InvokeElem(input);

                System.Collections.IList volList = output.
                    GetChildByName("volumes").GetChildren();
                System.Collections.IEnumerator volIter = volList.GetEnumerator();
                WriteObject("\n------------------------------------------------------------------------");
                WriteObject("Name \t type \t state \t size-total \t size-used \t size-available");
                WriteObject("------------------------------------------------------------------------");
                String vol = "";
                while (volIter.MoveNext())
                {
                    NaElement volInfo = (NaElement)volIter.Current;
                    vol = volInfo.GetChildContent("name") + "\t" + 
                        volInfo.GetChildContent("type") + "\t" + 
                        volInfo.GetChildContent("state") + "\t" + 
                        volInfo.GetChildContent("size-total") + "\t" + 
                        volInfo.GetChildContent("size-used") + "\t" + 
                        volInfo.GetChildContent("size-available");
                    WriteObject(vol);
                }
                WriteObject("------------------------------------------------------------------------\n");
            }
            catch (Exception e)
            {
                WriteObject(e.Message);
            }
        }
    }
}

File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/DotNet/VB.NET/vollist/vollist.vb
'======================================================================='
'                                                                       '
' vollist.vb                                                            '
'                                                                       '
'                                                                       '
' Sample code for using ONTAPI API to list volume information			'
'                                                                       '
' Copyright 2008 NetApp, Inc. All rights                                '
' reserved. Specifications subject to change without notice.            '
'                                                                       '
' This SDK sample code is provided AS IS, with no support or            '
' warranties of any kind, including but not limited to                  '
' warranties of merchantability or fitness of any kind,                 '
' expressed or implied.  This code is subject to the license            '
' agreement that accompanies the SDK.                                   '
'                                                                       '
' Usage: vollist <filer> <username> <passwd> [<volume>]                 '
'                                                                       '
'======================================================================='

Imports System
Imports System.Collections.Generic
Imports NetApp.Manage

Module VolList
    Public Sub Usage()
        Console.WriteLine(Environment.NewLine & "Usage: vollist <filer> <username> <passwd> [<volume>]")
    End Sub
    Sub Main()
        Dim args As String() = Environment.GetCommandLineArgs()

        If (args.Length < 4) Then
            Usage()
            Exit Sub
        End If

        Dim serverName As String = args(1)
        Dim user As String = args(2)
        Dim passwd As String = args(3)
        Dim volume As String = ""

        If (args.Length > 4) Then
            volume = args(4)
        End If

        Try
            'Initialize connection to server, and request version 1.0 of the API set.
            Dim server As New NaServer(serverName, 1, 0)

            server.ServerType = NaServer.SERVER_TYPE.FILER
            server.TransportType = NaServer.TRANSPORT_TYPE.HTTP
            server.SetAdminUser(user, passwd)

            'Frame the volume-list-Info ONTAPI API request
            Dim input As New NaElement("volume-list-info")
            If (volume <> "") Then
                input.AddNewChild("volume", volume)
            End If
            Dim output As NaElement

            'Invoke volume-list-Info ONTAPI API
            output = server.InvokeElem(input)

            'Get the list of children from 'output' element and iterate 
            'through each of the child element to retrieve their values
            Dim volList As List(Of NaElement) = output.GetChildByName("volumes").GetChildren()
            Dim total As Integer = volList.Count
            Dim volInfo As NaElement
            Dim index As Integer

            For index = 0 To total - 1
                Console.WriteLine("---------------------------------")
                volInfo = volList(index)
                Console.Write("Volume Name:")
                Console.WriteLine(volInfo.GetChildContent("name"))
                Console.Write("Volume State:")
                Console.WriteLine(volInfo.GetChildContent("state"))
                Console.Write("Disk Count:")
                Console.WriteLine(volInfo.GetChildIntValue("disk-count", -1))
                Console.Write("Total Files:")
                Console.WriteLine(volInfo.GetChildIntValue("files-total", -1))
                Console.Write("No of files used:")
                Console.WriteLine(volInfo.GetChildIntValue("files-used", -1))
            Next index
        Catch ex As NaException
            Console.WriteLine(ex.Message)
        Catch ex As Exception
            Console.WriteLine(ex.Message)
        End Try
    End Sub

End Module
